pkg/reconciler/workspace/controller.go

```diff
@@ -0,0 +1,385 @@
+	// Reconcile worked, nothing else to do for this workqueue item.
+	if err == nil {
+		klog.Infof("Successfully reconciled %q", key)
+		c.queue.Forget(key)
+		return
+	}
+
+	// Re-enqueue up to 5 times.
+	num := c.queue.NumRequeues(key)
+	if errors.IsRetryable(err) || num < 5 {
+		klog.Errorf("Error reconciling key %q, retrying... (#%d): %v", key, num, err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	// Give up and report error elsewhere.
+	c.queue.Forget(key)
+	runtime.HandleError(err)
+	klog.Errorf("Dropping key %q after failed retries: %v", key, err)
+}
+
+func (c *Controller) process(ctx context.Context, key string) error {
+	namespace, clusterAwareName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != "" {
+		return fmt.Errorf("namespace %q found in key for cluster-wide Workspace object", namespace)
+	}
+	clusterName, _ := clusters.SplitClusterAwareKey(clusterAwareName)
+
+	obj, err := c.workspaceLister.Get(key) // TODO: clients need a way to scope down the lister per-cluster
+	if err != nil {
+		return err
+	}
+	previous := obj.DeepCopy()
+
+	if err := c.reconcile(ctx, obj); err != nil {
+		return err
+	}
+
+	// If the object being reconciled changed as a result, update it.
+	if !equality.Semantic.DeepEqual(previous.Spec, obj.Spec) {
+		_, uerr := c.kcpClient.Cluster(clusterName).TenancyV1alpha1().Workspaces().Update(ctx, obj, metav1.UpdateOptions{})
+		return uerr
+	}
+	if !equality.Semantic.DeepEqual(previous.Status, obj.Status) {
+		_, uerr := c.kcpClient.Cluster(clusterName).TenancyV1alpha1().Workspaces().UpdateStatus(ctx, obj, metav1.UpdateOptions{})
+		return uerr
+	}
+
+	return nil
+}
+
+func (c *Controller) reconcile(ctx context.Context, obj *tenancyv1alpha1.Workspace) error {
+	if currentShard := obj.Status.Location.Current; currentShard != "" {
+		// make sure current shard still exists
+		shards, err := c.workspaceShardLister.List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		var found bool
+		for _, shard := range shards {
+			if shard.Name == currentShard {
+				found = true
+				break
+			}
+		}
+		if !found {
+			klog.Infof("removing spec.targetShard for workspace %q", obj.Name)
+			obj.Spec.TargetShard = ""
+		}
+	}
+	if obj.Spec.TargetShard == "" {
+		// find a shard for this workspace
+		shards, err := c.workspaceShardLister.List(labels.Everything())
+		if err != nil {
+			return err
+		}
+		if len(shards) != 0 {
+			target := shards[rand.Intn(len(shards))].Name
+			obj.Spec.TargetShard = target
+			obj.Status.Location.Target = target
+			klog.Infof("moving workspace %q to %q", obj.Name, target)
+		}
+	}
+	if obj.Spec.TargetShard != obj.Status.Location.Current {
+		klog.Infof("updating workspace %q status", obj.Name)
+		obj.Status.Location.Current = obj.Spec.TargetShard
+		obj.Status.Location.Target = ""
+	}
+	if obj.Status.Location.Current == "" {
+		obj.Status.Phase = tenancyv1alpha1.WorkspacePhaseInitializing
+		if !IsWorkspaceUnschedulable(obj) {
```

Comment:
If it's not unschedulable, then it's schedulable? Help me understand this flow?

Reply:
If it's not already unschedulable, make it so. I borrowed the conditions from Pod, and I also do not like the pattern but at least it has precedent? Open to other ideas for sur

Reply:
https://github.com/kcp-dev/kcp/pull/212#discussion_r752604413 will handle this